### ISSUE TITLE
Retrieving the manifest for an official DockerHub images (without the library prefix) causes a 500 error instead of a 401 error

### ISSUE THREAD

**Issue by matzegebbe** [distribution/distribution#3856]
We are using the registry as a pull-through cache, as described here: https://docs.docker.com/registry/recipes/mirror/, for our Renovate bot. The bot does HTTP calls for manifests with digests like http://docker-registry-mirror.kube-system.svc.cluster.local:5000/v2/openjdk/manifests/sha256:8e17383576d7e71988ee5927473a32e8461381c7a29eefa9a0c24b3a28926272. Since openjdk is an official Docker image, the vendor prefix (library) is missing. The HTTP response for that call is a 500 error with the following message:

```bash
curl -I http://localhost:5000/v2/openjdk/manifests/sha256:8e17383576d7e71988ee5927473a32e8461381c7a29eefa9a0c24b3a28926272 
-> HTTP/1.1 500 Internal Server Error
```
```json
{"errors":[{"code":"UNKNOWN","message":"unknown error","detail":{"errors":[{"code":"DENIED","message":"requested access to the resource is denied"},{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Action":"pull","Class":"","Name":"openjdk","Type":"repository"}]}]}}]}
```

doing the call with the /library/ return the manifest as expected:
```bash
curl -I http://localhost:5000/v2/library/openjdk/manifests/sha256:8e17383576d7e71988ee5927473a32e8461381c7a29eefa9a0c24b3a28926272
-> HTTP/1.1 200 OK
```
while the offical docker registry gives a 401

-  not working 401 
```bash
TOKEN=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:openjdk:pull" | jq -r .token)
curl -I -H "Authorization: Bearer ${TOKEN}" -H "Accept: application/vnd.docker.distribution.manifest.v2+json" -X GET https://registry-1.docker.io/v2/openjdk/manifests/sha256:8e17383576d7e71988ee5927473a32e8461381c7a29eefa9a0c24b3a28926272
```
- working 200
```bash
TOKEN=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:library/openjdk:pull" | jq -r .token)
curl -I -H "Authorization: Bearer ${TOKEN}" -H "Accept: application/vnd.docker.distribution.manifest.v2+json" -X GET https://registry-1.docker.io/v2/library/openjdk/manifests/sha256:8e17383576d7e71988ee5927473a32e8461381c7a29eefa9a0c24b3a28926272
```
**I think a 500er error is wrong in that case and the UNAUTHORIZED registry should return a 401**

error message in renovate:

```json
        "message": "Response code 500 (Internal Server Error)",
         "stack": "HTTPError: Response code 500 (Internal Server Error)\n    at Request.<anonymous> (/opt/buildpack/tools/renovate/34.157.1/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/34.157.1 (https://github.com/renovatebot/renovate)",
             "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "http://docker-registry-mirror.kube-system.svc.cluster.local:5000/v2/openjdk/manifests/sha256:8e17383576d7e71988ee5927473a32e8461381c7a29eefa9a0c24b3a28926272",
           "hostType": "docker",
           "username": "",
           "password": "",
           "method": "HEAD",
           "http2": false
         },
         "response": {
           "statusCode": 500,
           "statusMessage": "Internal Server Error",
           "body": "",
           "headers": {
             "content-type": "application/json; charset=utf-8",
             "docker-distribution-api-version": "registry/2.0",
             "x-content-type-options": "nosniff",
             "date": "Wed, 08 Mar 2023 11:06:54 GMT",
             "content-length": "290",
             "connection": "close"
           },
           "httpVersion": "1.1",
           "retryCount": 2
         }
```

**Comment by matzegebbe** [comment]
it still happens with `3.0.0-alpha.1`

```bash
docker run -it --rm -e REGISTRY_PROXY_REMOTEURL=https://registry-1.docker.io -p 5000:5000                                                                   ghcr.io/distribution/distribution:3.0.0-alpha.1
```

```bash
curl -I https://registry-1.docker.io/v2/openjdk/manifests/sha256:2775e341f3c17e6e3412c2edaa604f9d851f900757c7f193a1521bab8b4f46cb
# HTTP/1.1 401 Unauthorized
# body:
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Class":"","Name":"openjdk","Action":"pull"}]}]}
```

```bash
curl -I http://localhost:5000/v2/openjdk/manifests/sha256:2775e341f3c17e6e3412c2edaa604f9d851f900757c7f193a1521bab8b4f46cb
# HTTP/1.1 500 Internal Server Error
# body:
{"errors":[{"code":"UNKNOWN","message":"unknown error","detail":"errors:\ndenied: requested access to the resource is denied\nunauthorized: authentication required\n"}]}
```

This causes renovate to stop :/ 

body is the same response header not

**Comment by milosgajdos** [comment]
I have a strong suspicion that this has to do with the half-broken `client` code which we couldn't completely deprecate (other than shifting it to the `internal` package), because it's used by the proxy, rather than with distribution code itself.

My theory is that the client "swallows" the HTTP status code it receives from the remote registry which then results in proxy returning `500`. I will have to verify this, but this is my suspicion and I don't think there will be any fix possible without making a proper surgery on the client code which we really want to ditch (see: https://github.com/distribution/distribution?tab=readme-ov-file#how-does-this-integrate-with-docker-containerd-and-other-oci-client)

**Comment by milosgajdos** [comment]
Ok so, this is just for future reference but it goes kinda like this:

1. client attempts to fetch the manifest from remote registry https://github.com/distribution/distribution/blob/5f804a9df72f86e05de6a7c99466539b8fa26d2c/registry/proxy/proxymanifeststore.go#L51 -- it receives error (in this case it seems to be the `401` which you mention wrt to upstream)
2. The client then processes the response https://github.com/distribution/distribution/blob/5f804a9df72f86e05de6a7c99466539b8fa26d2c/internal/client/repository.go#L516 -- this is where the "trouble" happens
3. It turns out the token server responds with `insufficient_scopes` which the client maps to `403`: 
https://github.com/distribution/distribution/blob/5f804a9df72f86e05de6a7c99466539b8fa26d2c/internal/client/errors.go#L145-L146 -> https://github.com/distribution/distribution/blob/5f804a9df72f86e05de6a7c99466539b8fa26d2c/registry/api/errcode/register.go#L50-L56
4. none of that matters because we "swallow" the status code here by merging all the errors into a list of errors https://github.com/distribution/distribution/blob/5f804a9df72f86e05de6a7c99466539b8fa26d2c/internal/client/errors.go#L155
5. Which then ultimately ends up mapped into `500` CodeUnknown you're seeing https://github.com/distribution/distribution/blob/5f804a9df72f86e05de6a7c99466539b8fa26d2c/registry/handlers/manifests.go#L146

Like I said the fix is likely complicated, but feel free to take a stab at it 😄

**Comment by matzegebbe** [comment]
Thanks for checking so far. This is also what I see. We have subsequent errors that result in a 500 error, and the original 401 error gets swallowed.

![image](https://github.com/distribution/distribution/assets/3482021/fba69221-64b5-42e1-b1f5-40eb63aa2b9d)

**Comment by milosgajdos** [comment]
Yep. That client code was shehorned into proxy in a very terrible way, sadly.

**Comment by func0der** [comment]
@milosgajdos 

I am not quite sure I get the whole thing. With a configured Docker Hub Proxy, this is

What I experienced:

https://registry.example.com/v2/registry/postgres/manifests/17 --> broken (404), not 500 as mentioned here, but the response seems the same:

```
{"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":"unknown tag=17"}]}
```

https://registry.example.com/v2/registry/library/postgres/manifests/17 --> works and real manifest data is given out.

So.. Could we just add the `library` prefix for images coming along without any namespace?

Or did I miss the point here?

----

Edit:

Or should renovate normalize those images like dockers does with https://github.com/distribution/reference ?

**Comment by milosgajdos** [comment]
Every repo has a namespace or rather every repo must live in a namespace; what happens when you run `docker run nginx` is that Docker Hub normalizes that into `library/nginx` -- `library` is the Hub namespace for official images.

I am not sure it's a good idea to enforce this in the caching proxy; I'd say it likely isnt

**Comment by func0der** [comment]
So i guess this is a renovate thing to discuss then. I have started a [discussion over there](https://github.com/renovatebot/renovate/discussions/40402).
They are falling back to looking for `library/image` if `image` does not yield results already. But the only do this for the official docker registry. 

@matzegebbe Maybe the mentioned option 3 with [overridepackagename](https://docs.renovatebot.com/configuration-options/#overridepackagename) over there helps you out in the meantime. 

>  I am not sure it's a good idea to enforce this in the caching proxy; I'd say it likely isnt
Does this mean the issue is a "non-issue" here, since including behavior that is "unofficial" would cause more harm then good? Because I can totally get behind that :D
